### PR TITLE
Wasm `prepare()` doc

### PR DESCRIPTION
**Pull request by 0xisk**
Separating `prepare()` function in wasm page.

### CODE REVIEW

File: docs/prove/wasm.md [hunk 1]
```diff
@@ -24,4 +24,86 @@ This approach not only enhances privacy but also reduces network latency, as the
...
-TODO: add code snippets about preparing web workers and hooks that are basically the same for spartan and halo2.
+# First Step: Preparing the WebAssembly (Wasm) in the Browser
+
+For both the `halo2` and `spartan` proving systems, web workers function similarly at the ABI level. To begin importing and initializing the Wasm code in the browser, you should invoke worker.prepare(). This function handles the necessary steps to load and configure the Wasm module.
+
+## Error Reporting in the Browser
+
+When running Wasm code in the client-side browser, it's crucial to have effective error reporting to assist in debugging. The prepare function includes a call to `init_panic_hook()`, which sets up a panic hook. This hook redirects panic messages from Rust into the browser's console, making it easier to identify and resolve issues directly within the developer tools.
+
+This setup ensures that any runtime errors in the Rust code are clearly reported, with detailed stack traces available in the console, enhancing the debugging process during development.
```
Comment (sripwoud) [inline docs/prove/wasm.md:35]: haha good you are documenting this, I didn't know myself what the panic_hook was for this purpose!

File: docs/prove/wasm.md [hunk 2]
```diff
@@ -24,4 +24,86 @@ This approach not only enhances privacy but also reduces network latency, as the
 
 Next, we show an opiniated way (this is what our [UI](https://github.com/anonklub/anonklub/tree/main/ui) does) of wrapping these wasm and web workers features into React hooks for convenient use in a front end application:
 
-TODO: add code snippets about preparing web workers and hooks that are basically the same for spartan and halo2.
+# First Step: Preparing the WebAssembly (Wasm) in the Browser
+
+For both the `halo2` and `spartan` proving systems, web workers function similarly at the ABI level. To begin importing and initializing the Wasm code in the browser, you should invoke worker.prepare(). This function handles the necessary steps to load and configure the Wasm module.
```
Comment (sripwoud) [inline docs/prove/wasm.md:29]:
```suggestion
For both the `halo2` and `spartan` proving systems, web workers function similarly at the ABI level. To begin importing and initializing the Wasm code in the browser, you should invoke `worker.prepare()`. This function handles the necessary steps to load and configure the Wasm module.
```

File: docs/prove/wasm.md [hunk 3]
```diff
@@ -24,4 +24,86 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+
+## Error Reporting in the Browser
+
+When running Wasm code in the client-side browser, it's crucial to have effective error reporting to assist in debugging. The prepare function includes a call to `init_panic_hook()`, which sets up a panic hook. This hook redirects panic messages from Rust into the browser's console, making it easier to identify and resolve issues directly within the developer tools.
+
+This setup ensures that any runtime errors in the Rust code are clearly reported, with detailed stack traces available in the console, enhancing the debugging process during development.
+
+## Integration with Merkle Tree Packages
+
+The same prepare function is also available in the Merkle tree packages associated with each proving system: `@anonklub/halo2-binary-merkle-tree-worker` for the halo2 circuit and `@anonklub/merkle-tree-worker` for the spartan circuit. This ensures consistent Wasm initialization and error reporting across different components of the proving systems.
```
Comment (sripwoud) [inline docs/prove/wasm.md:39]:
```suggestion
The same `prepare` function is also available in the Merkle tree packages associated with each proving system: `@anonklub/halo2-binary-merkle-tree-worker` for the halo2 circuit and `@anonklub/merkle-tree-worker` for the spartan circuit. This ensures consistent Wasm initialization and error reporting across different components of the proving systems.
```

File: docs/prove/wasm.md [hunk 4]
```diff
@@ -24,4 +24,86 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+
+## Thread Pool Initialization in Halo2 Packages
+
+The `halo2` packages include an additional step in their `prepare` functions that is not present in the `spartan` packages. Specifically, the `halo2` packages use [wasm-bindgen-rayon](https://github.com/RReverser/wasm-bindgen-rayon) to enable multi-threading within the WebAssembly environment.
+
+As part of this setup, the `prepare` function checks the number of hardware threads available on the user's device via `navigator.hardwareConcurrency`. This information is then used to initialize a thread pool using `initThreadPool(numThreads)`, which allows the WebAssembly code to efficiently utilize multiple threads for parallel processing tasks.
+
+## Implementation Details of the prepare Function
```
Comment (sripwoud) [inline docs/prove/wasm.md:47]:
```suggestion
## Implementation Details of the `prepare` Function
```

File: docs/prove/wasm.md [hunk 5]
```diff
@@ -24,4 +24,86 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+
+Each package has its own specific implementation of the prepare function, tailored to its particular needs:
```
Comment (sripwoud) [inline docs/prove/wasm.md:49]:
```suggestion
Each package has its own specific implementation of the `prepare` function, tailored to its particular needs:
```

**Review by sripwoud — CHANGES_REQUESTED** [review]
overall ACK

Just left some suggestions to wrap `prepare` in backticks everywhere

**Comment by 0xisk** [comment]
> overall ACK
> 
> Just small requests for changes:
> 
>     * I suggest wrapping `prepare` in backticks everywhere
> 
>     * can you add a part about using the web workers in react with our custom `useWorker` hook
>       e.g we can write something like this no (the same `useWorker` hook works for all workers right)? Then just refer/link to that part in the necessary section of the halo2 and spartan pages.
>       ```ts
>       import { SpartanEcdsaWorker } from '@anonklub/spartan-ecdsa-worker'
>       import { MerkleTreeWorker } from '@anonklub/merkle-tree-worker'
>       import { Halo2EthMembershipWorker } from '@anonklub/halo2-eth-membership-worker'
>       
>       type Worker =
>         | typeof SpartanEcdsaWorker
>         | typeof MerkleTreeWorker
>         | typeof Halo2EthMembershipWorker
>       
>       export const useWorker = (worker: Worker) => {
>         const [isWorkerReady, setIsWorkerReady] = useState(false)
>       
>         useEffect(() => {
>           void (async () => {
>             await worker.prepare()
>             setIsWorkerReady(true)
>           })()
>         }, [])
>       
>         return isWorkerReady
>       }
>       ```

That is updated, could you check please?

File: docs/prove/wasm.md [hunk 6]
```diff
@@ -24,4 +24,123 @@ This approach not only enhances privacy but also reduces network latency, as the
 
 Next, we show an opiniated way (this is what our [UI](https://github.com/anonklub/anonklub/tree/main/ui) does) of wrapping these wasm and web workers features into React hooks for convenient use in a front end application:
 
-TODO: add code snippets about preparing web workers and hooks that are basically the same for spartan and halo2.
+# First Step: Preparing the WebAssembly (Wasm) in the Browser
+
+For both the `halo2` and `spartan` proving systems, web workers function similarly at the ABI level. To begin importing and initializing the Wasm code in the browser, you should invoke worker.prepare(). This function handles the necessary steps to load and configure the Wasm module.
```
Comment (sripwoud) [inline docs/prove/wasm.md:29]:
```suggestion
For both the `halo2` and `spartan` proving systems, web workers function similarly at the ABI level. To begin importing and initializing the Wasm code in the browser, you should invoke `worker.prepare()`. This function handles the necessary steps to load and configure the Wasm module.
```

File: docs/prove/wasm.md [hunk 7]
```diff
@@ -24,4 +24,123 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+
+As part of this setup, the `prepare()` function checks the number of hardware threads available on the user's device via `navigator.hardwareConcurrency`. This information is then used to initialize a thread pool using `initThreadPool(numThreads)`, which allows the WebAssembly code to efficiently utilize multiple threads for parallel processing tasks.
+
+## Implementation Details of the `prepare()` Function
+
+Each package has its own specific implementation of the `prepare()` function, tailored to its particular needs:
+
+1. Merkle Tree Web Worker (@anonklub/merkle-tree-worker):
+
+```ts
```
Comment (sripwoud) [inline docs/prove/wasm.md:53]:
```suggestion
```js
```

i know it is ts code, but docsify does not support a specific parser for typescript, so this code block will be rendered without syntax highlighting.
If we use `js` then we will have syntax highlighting.

image
vs
image

File: docs/prove/wasm.md [hunk 8]
```diff
@@ -24,4 +24,123 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+async prepare() {
+    merkleTreeWasm = await import('@anonklub/merkle-tree-wasm');
+}
+```
+
+2. Spartan Circuit Web Worker (@anonklub/spartan-ecdsa-worker):
+
+```ts
```
Comment (sripwoud) [inline docs/prove/wasm.md:61]:
```suggestion
```js
```

File: docs/prove/wasm.md [hunk 9]
```diff
@@ -24,4 +24,123 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+    if (!initialized) {
+        spartanEcdsaWasm.prepare();
+        initialized = true;
+    }
+}
+```
+
+3. Halo2 Binary Merkle Tree Web Worker (@anonklub/halo2-binary-merkle-tree-worker):
+
+```ts
```
Comment (sripwoud) [inline docs/prove/wasm.md:75]:
```suggestion
```js
```

File: docs/prove/wasm.md [hunk 10]
```diff
@@ -24,4 +24,123 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+    await halo2BinaryMerkleTreeWasm.initSync(bufferSource);
+    await halo2BinaryMerkleTreeWasm.initPanicHook();
+    const numThreads = navigator.hardwareConcurrency;
+    await halo2BinaryMerkleTreeWasm.initThreadPool(numThreads);
+}
+```
+
+4. Halo2 Circuit Web Worker (@anonklub/halo2-eth-membership):
+
+```ts
```
Comment (sripwoud) [inline docs/prove/wasm.md:92]:
```suggestion
```js
```

File: docs/prove/wasm.md [hunk 11]
```diff
@@ -24,4 +24,123 @@ This approach not only enhances privacy but also reduces network latency, as the
...
+
+## Creating a Simple React Hook
+
+To make sure the Wasm module is ready before your app uses it, you can create a hook called `useWorker`. This hook will load the worker and set it up when your component mounts.
+
+### Example: `useWorker` Hook
+
+Here’s a simple example of how to create and use the `useWorker` hook:
+
+```ts
```
Comment (sripwoud) [inline docs/prove/wasm.md:123]:
```suggestion
```js
```

**Review by sripwoud — CHANGES_REQUESTED** [review]
some final request for changes about the formatting. (you can check how it renders with `pnpm serve`)
Content is good :+1:

File: docs/prove/wasm.md (hunk 7 above)
Comment (0xisk) [reply docs/prove/wasm.md:53]: I am able to see syntax highlighted in this doc for example: https://github.com/anonklub/docs/blob/docs/wasm/docs/prove/wasm.md
Is it not shown in your end?!

Comment (0xisk) [reply docs/prove/wasm.md:53]: My bad I didn't think of that.

Comment (0xisk) [reply docs/prove/wasm.md]: Done.

**Review by sripwoud — APPROVED** [review]
ACK

**Pull request merged**